python-package/lightgbm/dask.py

```diff
@@ -134,22 +98,27 @@ def _assign_open_ports_to_workers(
     worker_to_port: dict
         mapping from worker address to an open port.
     """
-    host_ports_futures = {}
-    for hostname, workers in host_to_workers.items():
-        n_workers_in_host = len(workers.all_workers)
-        host_ports_futures[hostname] = client.submit(
-            _find_n_open_ports,
-            n=n_workers_in_host,
-            workers=[workers.default],
-            pure=False,
+    # Acquire port in worker
+    worker_to_future = {}
+    for worker in workers:
+        worker_to_future[worker] = client.submit(
+            _acquire_port,
+            workers=[worker],
             allow_other_workers=False,
+            pure=False,
         )
-    found_ports = client.gather(host_ports_futures)
-    worker_to_port = {}
-    for hostname, workers in host_to_workers.items():
-        for worker, port in zip(workers.all_workers, found_ports[hostname]):
-            worker_to_port[worker] = port
-    return worker_to_port
+
+    # schedule futures to retrieve each element of the tuple
+    worker_to_socket_future = {}
+    worker_to_port_future = {}
+    for worker, socket_future in worker_to_future.items():
+        worker_to_socket_future[worker] = client.submit(operator.itemgetter(0), socket_future)
+        worker_to_port_future[worker] = client.submit(operator.itemgetter(1), socket_future)
+
+    # retrieve ports
+    worker_to_port = client.gather(worker_to_port_future)
```

Comment:
This is the most important line, and I don't fully understand how it'll work.

If each `_RemoteSocket` contains a `socket.socket` in its attributes, and `socket.socket` attributes aren't pickleable with `cloudpickle`... 

```python
import cloudpickle
import socket

s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind(('', 0))

port = s.getsockname()[1]
print(f"port: {port}")

with open("out.pkl", "wb") as f:
    cloudpickle.dump(s, f)

# TypeError: cannot pickle 'socket' object
```

... how can an object created by a run of `_acquire_port()` on a worker be collected back to the client?

And even if it were... how does this work in a non-`LocalCluster` cluster, where each worker exists in the process space of a different machine?

Reply:
That's why we return a tuple from the function and only collect the second element (the integer containing the port number). The first element (which contains the socket) stays on the worker, we just keep the future and then use it when calling the train function, so it only gets used in its local worker.

It's kind of like an actor, we're creating the object on the remote worker and only keeping a reference to it, we don't ever ask for it in the client, so it never has to be serialized.

Reply:
oooooooo interesting! I guess it's the combination of `workers=[worker]` and `pure=False` that prevents Dask from ever trying to move that result (the tuple with the `_RemoteSocket` and `port`) over the network? That's very clever, @jmoralez 😁 

I support it!